### PR TITLE
Config Improvements

### PR DESCRIPTION
**Pull request by zramsay**
most of #628 
refs: #709, #733

### CODE REVIEW

**Comment by melekes** [comment]
Tests failed with:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x902469]

goroutine 41 [running]:
github.com/tendermint/tendermint/rpc/lib/client.callWgDoneOnResult(0xc4201ce3c0, 0xc42009f7a0, 0xc42019d870)
        /go/src/github.com/tendermint/tendermint/rpc/lib/client/ws_client_test.go:207 +0x1a9
created by github.com/tendermint/tendermint/rpc/lib/client.TestWSClientReconnectsAfterWriteFailure
        /go/src/github.com/tendermint/tendermint/rpc/lib/client/ws_client_test.go:100 +0x192
exit status 2
FAIL    github.com/tendermint/tendermint/rpc/lib/client 2.339s
```

but this seems to be not related. I will take a look

**Comment by melekes** [comment]
```
test/app/test.sh: line 80: 13549 Killed                  abci-cli counter --serial --abci grpc > /dev/null
test/app/test.sh: line 80: 13550 Killed                  tendermint node --abci grpc > tendermint.log
Starting counter_over_grpc_grpc (ie. with grpc broadcast_tx)
running test
... sending tx. expect no error
... building grpc_client
RESPONSE
rpc error: code = Unavailable desc = grpc: the connection is unavailable
parse error: Invalid numeric literal at line 1, column 4
parse error: Invalid numeric literal at line 1, column 4
-------
TX 00
RESPONSE rpc error: code = Unavailable desc = grpc: the connection is unavailable
ERROR rpc error: code = Unavailable desc = grpc: the connection is unavailable
----
Unexpected error sending tx (00): rpc error: code = Unavailable desc = grpc: the connection is unavailable
```

**Comment by zramsay** [comment]
hmm, i ran `./test/app/test.sh` and got
``` 
Running all
Starting dummy_over_socket
running test
0x616263643D64636261
```

**Comment by melekes** [comment]
The tests behave themselves very strange. Every time I am rebuilding this, they fail with a new error.

File: config/config.go [hunk 1]
```diff
@@ -365,7 +384,7 @@ func (cfg *ConsensusConfig) PeerQueryMaj23Sleep() time.Duration {
 // DefaultConsensusConfig returns a default configuration for the consensus service
 func DefaultConsensusConfig() *ConsensusConfig {
 	return &ConsensusConfig{
-		WalPath:                     "data/cs.wal/wal",
+		WalPath:                     filepath.Join(defaultDataDir, "cs.wal/wal"),
```
Comment (melekes) [inline config/config.go]: `filepath.Join(defaultDataDir, "cs.wal", "wal")`

File: config/toml.go [hunk 2]
```diff
@@ -1,54 +1,194 @@
...
+
+# TCP or UNIX socket address for the profiling server to listen on
+prof_laddr = "{{ .BaseConfig.ProfListenAddress }}"
+
+# If true, query the ABCI app on connecting to a new peer
+# so the app can decide if we should keep the connection or not
+filter_peers = {{ .BaseConfig.FilterPeers }}
+
+# What indexer to use for transactions
+tx_index = "{{ .BaseConfig.TxIndex }}"
```
Comment (melekes) [inline config/toml.go]: this moved to its own section `[tx_index]`

File: docs/specification/configuration.rst [hunk 3]
```diff
@@ -1,57 +1,151 @@
...
+    
+    # TCP or UNIX socket address for the profiling server to listen on
+    prof_laddr = ""
+    
+    # If true, query the ABCI app on connecting to a new peer
+    # so the app can decide if we should keep the connection or not
+    filter_peers = false
+    
+    # What indexer to use for transactions
+    tx_index = "kv"
```
Comment (melekes) [inline docs/specification/configuration.rst]: same

**Comment by melekes** [comment]
```
2017-12-02T05:48:17Z box1393.localdomain docker/local_testnet_1[14895]: ERROR: Failed to create node: Couldn't read GenesisDoc file: open /go/src/github.com/tendermint/tendermint/test/p2p/data/mach1/core/config/genesis.json: no such file or directory
```

**Comment by melekes** [comment]
also, dockerfile needs to be updated here

**Comment by zramsay** [comment]
after discusssion with bucky, we need to not break this, and provide an upgrade path with a deprecation warning

**Comment by melekes** [comment]
```
# github.com/tendermint/tendermint/config
config/toml.go:38:44: undefined: defaultConfig
config/toml.go:244:44: undefined: testConfig
```

**Comment by zramsay** [comment]
on it

File: docs/specification/configuration.rst [hunk 4]
```diff
@@ -1,58 +1,170 @@
...
--  ``rpc.grpc_laddr``: GRPC listen address (BroadcastTx only). Port
-   required. *Default*: ``""``
--  ``rpc.laddr``: RPC listen address. Port required. *Default*:
-   ``"0.0.0.0:46657"``
--  ``rpc.unsafe``: Enabled unsafe rpc methods. *Default*: ``true``
+Tendermint Core can be configured via a TOML file in
+``$TMHOME/config/config.toml``. Some of these parameters can be overridden by
+command-line flags. For most users, the options in the ``##### main
+base configuration options #####`` are intended to be modified while
+config options further below are intended for advance power users.
```
Comment (ebuchman) [inline docs/specification/configuration.rst]: not quite true. theres stuff like `p2p.seeds` that the average user should be manipualting

**Comment by zramsay** [comment]
currently hangs in the CI

**Comment by zramsay** [comment]
narrowed down the test failure when rebasing this branch on the `jenkins` branch: https://ci.interblock.io/job/05.Test.Tendermint.FeatureBranches.p2p/98/console
see the branch `config-on-jenkins`

**Comment by zramsay** [comment]
same error when running locally

**Comment by zramsay** [comment]
boom: fixed with latest commit https://ci.interblock.io/job/05.Test.Tendermint.FeatureBranches.p2p/100/console

**Pull request merged**